Barber-Agakov, and OED refactoring

ae-foster opened the pull request:
New:
 - Barber-Agakov estimator of EIG
 - models and guides for sigmoid link function linear model

Refactored:
 - consistent API for different EIG estimators
 - improved code re-use in guides
 - extended support of EIG estimators: include integrating out random effects, supporting lists of variable names as well as single names

eb8680 commented:
@ae-foster to simplify reviewing, if some of this code is from old PRs merged to `oed-master`, can you add links to those PRs next to the appropriate bullet points above?

ae-foster commented:
Review of incorporated PRs:
 - https://github.com/uber/pyro/pull/1326/files smartened up the EIG estimator API (making it consistent between different functions). Moved the 'optimise a loss function' code from benchmarking script to `eig.py`. There are also a bunch of redundant changes to guide. I would only look at the `eig.py` diff
 - https://github.com/uber/pyro/pull/1331/files added sigmoid model and looks like some further refactors related to integrating out nuisance variables. Need only look at the diff for `bayes_linear.py` and `eig.py`
 - https://github.com/uber/pyro/pull/1342 brought the BA guides to their current form (also added linalg changes already merged to `dev` in https://github.com/uber/pyro/pull/1348). Probably only need to look at `amort.py` -- new file.

Overall, you could skip reading changes for `eig_estimation_benchmarking.py`, `ab_test.py` and `item_response.py`. There are no substantial changes- just matching new API.

ae-foster commented on examples/contrib/oed/eig_estimation_benchmarking.py:
```diff
@@ -7,148 +7,383 @@
-
-def naive_rainforth_lm(X_lm, w_sds, N, M):
-    model, _ = zero_mean_unit_obs_sd_lm(w_sds)
-    return naive_rainforth(model, X_lm, "y", "w", N=N, M=M)
```
Refactor of `eig.py` means these wrapper functions are no longer necessary

ae-foster commented on examples/contrib/oed/eig_estimation_benchmarking.py:
```diff
@@ -7,148 +7,383 @@
-        opt(params)
-    _, dv_loss = dv_loss_fn(final_X, final_n_samples)
-    if return_history:
-        return torch.stack(history), dv_loss
```
Moved to `eig.py` and refactored to support DV and BA together

ae-foster commented on examples/contrib/oed/eig_estimation_benchmarking.py:
```diff
@@ -7,148 +7,383 @@
+    trace = pyro.poutine.trace(model).get_trace(expanded_design)
+    trace.compute_log_prob()
+    lp = sum(trace.nodes[l]["log_prob"] for l in target_labels)
+    return -lp.sum(0)/num_samples
```
These prior entropy estimators are used to switch between APE and EIG for plotting the final graphs. One uses lm formula, the other uses Monte Carlo

ae-foster commented on examples/contrib/oed/eig_estimation_benchmarking.py:
```diff
@@ -7,148 +7,383 @@
-      (X_lm, naive_rainforth_lm, [torch.tensor([10., .1]), 2000, 2000]),
-      (X_lm, donsker_varadhan_lm, [torch.tensor([10., .1]), 4000, 200, 0.005, T_neural(2, 2)])],
+        hprior = mc_H_prior(model, design, observation_labels, target_labels)
+    return hprior - barber_agakov_ape(model, design, observation_labels, target_labels, *args, **kwargs)
```
These wrapper functions then use the prior entropy estimators

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+        tikhonov_diag = torch.diag(self.softplus(self.tikhonov_diag))
+        xtx = torch.matmul(design.transpose(-1, -2), design) + tikhonov_diag
+        xtxi = rinverse(xtx, sym=True)
+        mu = rmv(xtxi, rmv(design.transpose(-1, -2), y))
```
Implements `mu_hat = (X^T X + Lambda)^{-1} X^T Y` where Lambda is a `nn.Parameter`. This is tikhonov regularisation (hence the name). You could actually calculate Lambda from the prior and emission covariances

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+
+        # Extract sub-indices
+        mu = tensor_to_dict(self.w_sizes, mu, subset=target_labels)
+        scale_tril = {l: rtril(self.scale_tril[l]) for l in target_labels}
```
Assumption implicit: `scale_tril` is not dependent upon `y`

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+
+        for l in target_labels:
+            w_dist = dist.MultivariateNormal(mu[l], scale_tril=scale_tril[l])
+            pyro.sample(l, w_dist)
```
Flaw: does not account for correlation between differently named variables (does not affect any existing example though)

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+        hidden = self.softplus(y_trans)
+        y_trans = y_trans + hidden * self.h1_weight + self.h1_bias
+
+        return self.linear_model_formula(y_trans, design, target_labels)
```
Based on the doc I wrote up, we can use the linear model formula replacing `y` with its expectation conditional on `\tilde{y}`. We ought to add a `y` dependent covariance term as well (missing for now)

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+
+        yty = rvv(y, y)
+        ytxmu = rvv(y, rmv(design, mu_vec))
+        beta = self.b0 + .5*(yty - ytxmu)
```
Direct application of formula

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+        ytxmu = rvv(y, rmv(design, mu_vec))
+        beta = self.b0 + .5*(yty - ytxmu)
+
+        return mu, scale_tril, self.alpha, beta
```
`alpha` does not depend on `y` so can make a parameter

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+
+        for label in target_labels:
+            if label != self.tau_label:
+                if self.mf:
```
If mean field, posterior is Gaussian and independent of `tau`. Otherwise they are coupled in the correct NIG way

ae-foster replied:
Not clear what would happen if tau was not a variable of interest but you wanted non-mean-field

ae-foster commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,152 @@
+                pyro.sample(label, w_dist)
+
+
+class GuideDV(nn.Module):
```
Add docstring

ae-foster commented on examples/contrib/oed/models/bayes_linear.py:
```diff
@@ -1,15 +1,126 @@
+    # For testing, add these
+    model.obs_sd = observation_sd
+    model.w_sds = {coef_label: coef_sd}
+    return model
```
These are just utility functions calling `bayesian_linear_model` with different combinations of arguments

ae-foster commented on examples/contrib/oed/models/bayes_linear.py:
```diff
@@ -1,15 +1,126 @@
+def normal_guide(observation_sd, coef_shape, coef_label="w"):
+    return partial(normal_inv_gamma_family_guide,
+                   obs_sd=observation_sd,
+                   w_sizes={coef_label: coef_shape})
```
Ditto for `normal_inv_gamma_family_guide`

ae-foster commented on examples/contrib/oed/models/bayes_linear.py:
```diff
@@ -1,15 +1,126 @@
+        k_shape = batch_shape + (sigmoid_design.shape[-1],)
+        k = pyro.sample(sigmoid_label,
+                        dist.Gamma(sigmoid_alpha.expand(k_shape),
+                                   sigmoid_beta.expand(k_shape)).independent(1))
```
In this sigmoid model, the random slope has a Gamma distribution. Could easily add a different one and use `bayesian_linear_model` to do the rest

ae-foster commented on examples/contrib/oed/models/bayes_linear.py:
```diff
@@ -67,47 +191,76 @@ def bayesian_linear_model(design, w_means=None, w_sqrtlambdas=None, obs_sd=None,
+        group_p = alpha.shape[-1]
+        G_prior = dist.Gamma(alpha.expand(tau_shape + (group_p,)),
+                             beta.expand(tau_shape + (group_p,)))
+        G = 1./torch.sqrt(pyro.sample("G_" + name, G_prior))
```
These parameters have a shared diagonal covariance matrix `G`

ae-foster commented on examples/contrib/oed/models/bayes_linear.py:
```diff
@@ -67,47 +191,76 @@ def bayesian_linear_model(design, w_means=None, w_sqrtlambdas=None, obs_sd=None,
-def normal_inv_gamma_guide(design, obs_sd, w_sizes):
+# TODO replace this guide with one allowing correlation between
+# regression coefficients (necessary!)
+def normal_inv_gamma_family_guide(design, obs_sd, w_sizes, mf=False):
```
This is flagged for rewriting because it ignores all correlations between coefficients

ae-foster commented on examples/contrib/oed/models/bayes_linear.py:
```diff
@@ -119,80 +272,18 @@ def normal_inv_gamma_guide(design, obs_sd, w_sizes):
-
-
-# Some wrapper functions for common models
-def zero_mean_unit_obs_sd_lm(prior_sds, intercept_sd=None):
```
This code just moved to the top of the file :)

ae-foster commented on pyro/contrib/autoguide/__init__.py:
```diff
@@ -756,19 +756,20 @@ def __call__(self, *args, **kwargs):
     for name, site in trace.nodes.items():
         if site["type"] == "sample":
             if not poutine.util.site_is_subsample(site):
-                # TODO: optionally check pattern match here
```
Done this using `whitelist`

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -6,10 +6,11 @@
-def vi_ape(model, design, observation_labels, vi_parameters, is_parameters,
-           y_dist=None):
+def vi_ape(model, design, observation_labels, target_labels,
+           vi_parameters, is_parameters, y_dist=None):
```
The common API for eig estimators is `eig(model, design, observation_labels, target_labels, ...)`

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -55,7 +60,7 @@ def posterior_entropy(y_dist, design):
         SVI(conditioned_model, **vi_parameters).run(design)
         # Recover the entropy
-        return mean_field_guide_entropy(vi_parameters["guide"], design)  # TODO specify inference labels pattern here
+        return mean_field_guide_entropy(vi_parameters["guide"], [design], whitelist=target_labels)
```
You can pick up entropy from only some sites now. Require mean field (as before)

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -68,61 +73,201 @@ def posterior_entropy(y_dist, design):
+        entropy is to be measured.
+    :param int N: Number of outer expectation samples.
+    :param int M: Number of inner expectation samples for `p(y|d)`.
+    :param int M_prime: Number of samples for `p(y | theta, d)` if required.
```
New: can integrate out random effects with rainforth by using nested MC on both terms

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -68,61 +73,201 @@ def posterior_entropy(y_dist, design):
+    if isinstance(target_labels, str):
+        target_labels = [target_labels]
+    loss = donsker_varadhan_loss(model, T, observation_labels, target_labels)
+    return opt_eig_ape_loss(design, loss, num_samples, num_steps, optim, return_history,
```
Just calls the `opt_eig_ape_loss` function with the right loss. This allows re-use of `opt` between DV and BA

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -133,20 +278,21 @@ def donsker_varadhan_loss(model, observation_label, T):
         # Unshuffled data
         unshuffled_trace = poutine.trace(model).get_trace(expanded_design)
-        y = unshuffled_trace.nodes[observation_label]["value"]
+        y_dict = {l: unshuffled_trace.nodes[l]["value"] for l in observation_labels}
```
Better support when we have a list of variables not just one

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -162,6 +308,30 @@ def loss_fn(design, num_particles):
     return loss_fn
 
 
+def barber_agakov_loss(model, guide, observation_labels, target_labels):
```
Actual implementation of BA!

ae-foster commented on tests/contrib/autoguide/test_mean_field_guide_entropy.py:
```diff
@@ -25,4 +25,4 @@ def h(p):
 ])
 def test_guide_entropy(guide, args, expected_entropy):
-    assert_equal(mean_field_guide_entropy(guide, *args), expected_entropy)
+    assert_equal(mean_field_guide_entropy(guide, args), expected_entropy)
```
An annoying Python 2-ism

ae-foster commented:
End-to-end EIG estimation tests now on branch https://github.com/uber/pyro/tree/barber-agakov-tests

martinjankowiak commented:
@ae-foster let's try to get this merged today. i think you need to merge #1369 to get CI to be happy.

ae-foster commented:
I'll try and find some time to go over this in person with @eb8680 (unless he's cool to just merge). Ok I'll rebase from dev to get the pytest pin

eb8680 commented on examples/contrib/oed/eig_estimation_benchmarking.py:
```diff
@@ -7,148 +9,383 @@
+
+
+@pytest.mark.parametrize("title,model,design,observation_label,target_label,arglist", [
+    ("Sigmoid link function",
```
Move these test cases into a list of tuples so they're easier to read; see e.g. [HMC tests](https://github.com/uber/pyro/blob/dev/tests/infer/mcmc/test_hmc.py#L60)

eb8680 commented on examples/contrib/oed/eig_estimation_benchmarking.py:
```diff
@@ -7,148 +9,383 @@
-     (X_small, torch.tensor([10., 2.5]), {"n_iter": 2000, "n_samples": 2000, "lr": 0.0005, "T": T_neural(2, 2),
-                                          "final_n_samples": 10000}),
+@pytest.mark.parametrize("title,model,design,observation_label,target_label,est1,est2,kwargs1,kwargs2", [
+    ("Barber-Agakov on difficult sigmoid",
```
Move to list of tuples as in e.g. [HMC tests](https://github.com/uber/pyro/blob/dev/tests/infer/mcmc/test_hmc.py#L60)

eb8680 commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,161 @@
+        y = torch.cat(list(y_dict.values()), dim=-1)
+
+        # Approx invert transformation on y in expectation
+        y, y1m = y.clamp(1e-35, 1), (1.-y).clamp(1e-35, 1)
```
Double check that this is stable on CUDA

eb8680 commented on examples/contrib/oed/guides/amort.py:
```diff
@@ -0,0 +1,161 @@
+
+        return mu, scale_tril
+
+    def guide(self, y_dict, design, observation_labels, target_labels):
```
nit: make this the `forward` method of the `Module`
